src/main/scala-2.13+/scala/concurrent/stm/TMap.scala

```diff
@@ -0,0 +1,127 @@
+
+  /** Constructs and returns a new `TMap` that will contain the key/value pairs
+   *  from `it`.
+   */
+  def from[A, B](it: IterableOnce[(A, B)]): TMap[A, B] = {
+    val b = TMap.newBuilder[A, B]
+    val sizeHint = it.knownSize
+    if (sizeHint >= 0)
+      b.sizeHint(sizeHint)
+    b ++= it
+    b.result()
+  }
+
+
+  /** Allows a `TMap` in a transactional context to be used as a `Map`. */
+  implicit def asMap[A, B](m: TMap[A, B])(implicit txn: InTxn): View[A, B] = m.single
+}
+
+
+/** A transactional map implementation that requires that all of its map-like
+ *  operations be called from inside an atomic block.  Rather than extending
+ *  `Map`, an implicit conversion is provided from `TMap` to `Map` if the
+ *  current scope is part of an atomic block (see `TMap.asMap`).
+ *
+ *  The keys (with type `A`) must be immutable, or at least not modified while
+ *  they are in the map.  The `TMap` implementation assumes that it can safely
+ *  perform key equality and hash checks outside a transaction without
+ *  affecting atomicity.
+ *
+ *  @author Nathan Bronson
+ */
+trait TMap[A, B] extends TxnDebuggable {
+
+  /** Returns an instance that provides transactional map functionality without
+   *  requiring that operations be performed inside the static scope of an
+   *  atomic block.
+   */
+  def single: TMap.View[A, B]
+
+  def clone(implicit txn: InTxn): TMap[A, B] = single.clone.tmap
+
+  // The following method work fine via the asMap mechanism, but is important
+  // enough that we don't want the implicit conversion to make it invisible to
+  // ScalaDoc or IDE auto-completion
+
+  def snapshot: immutable.Map[A, B] = single.snapshot
+
+  // The following methods work fine via the asMap mechanism, but are heavily
+  // used.  We add transactional versions of them to allow overrides to get
+  // access to the InTxn instance without a ThreadLocal lookup.
+
+  def isEmpty(implicit txn: InTxn): Boolean
+  def size(implicit txn: InTxn): Int
+  def foreach[U](f: ((A, B)) => U)(implicit txn: InTxn): Unit
+  def contains(key: A)(implicit txn: InTxn): Boolean
+  def apply(key: A)(implicit txn: InTxn): B
+  def get(key: A)(implicit txn: InTxn): Option[B]
+  def update(key: A, value: B)(implicit txn: InTxn): Unit = put(key, value)
+  def put(key: A, value: B)(implicit txn: InTxn): Option[B]
+  def remove(key: A)(implicit txn: InTxn): Option[B]
+
+  // The following methods return the wrong receiver when invoked via the asMap
+  // conversion.  They are exactly the methods of mutable.Map whose return type
+  // is this.type.  Note that there are other methods of mutable.Map that we
+  // allow to use the implicit mechanism, such as getOrElseUpdate(k).
+
+  def += (kv: (A, B))(implicit txn: InTxn): this.type = { put(kv._1, kv._2) ; this }
+  def += (kv1: (A, B), kv2: (A, B), kvs: (A, B)*)(implicit txn: InTxn): this.type = { this += kv1 += kv2 ++= kvs }
```

Comment:
FTR, this method has been deprecated in 2.13, in favor of `++=`. I think we should follow the same convention here.

Reply:
The += on `Growable` is not deprecated: https://github.com/scala/scala/blob/d32c9efd102a777b0bb4099b78d14d6f1ae42b92/src/library/scala/collection/mutable/Growable.scala#L39 is this an oversight?

Reply:
There is an unmerged PR for that: https://github.com/scala/scala/pull/6719